packages/flutter/lib/src/widgets/animated_child_switcher.dart

```diff
@@ -10,39 +10,135 @@ import 'framework.dart';
 
+// Internal representation of a child that, now or in the past, was set on the
+// AnimatedChildSwitcher.child field, but is now in the process of
+// transitioning. The internal representation includes fields that we don't want
+// to expose to the public API (like the controller).
 class _AnimatedChildSwitcherChildEntry {
-  _AnimatedChildSwitcherChildEntry(this.widget, this.controller, this.animation);
+  _AnimatedChildSwitcherChildEntry({
+    @required this.widgetChild,
+    @required this.transition,
+    @required this.animation,
+    @required this.controller,
+  })  : assert(transition != null),
+        assert(animation != null),
+        assert(controller != null);
 
-  Widget widget;
+  final Animation<double> animation;
+
+  // The currently built transition for this child.
+  Widget transition;
 
+  // The animation controller for the child's transition.
   final AnimationController controller;
-  final Animation<double> animation;
+
+  // The widget's child at the time this entry was created or updated.
+  Widget widgetChild;
 }
 
-/// A widget that automatically does a [FadeTransition] between a new widget and
+/// Signature for builders used to generate custom transitions for
+/// [AnimatedChildSwitcher].
+///
+/// The [child] should be transitioned in when [animation] is running in the
+/// forward direction,
+///
+/// The function should return a widget which wraps the given [child]. It may
+/// also use the [animation] to inform its transition. It must not return null.
+typedef Widget AnimatedChildSwitcherTransitionBuilder(Widget child, Animation<double> animation);
+
+/// Signature for builders used to generate custom layouts for
+/// [AnimatedChildSwitcher].
+///
+/// The [children] should be positioned according to their corresponding
+/// animation.
+///
+/// The function should return a widget which contains the given children, laid
+/// out as desired. It must not return null.
+typedef Widget AnimatedChildSwitcherLayoutBuilder(List<Widget> children);
+
+/// A widget that by default does a [FadeTransition] between a new widget and
 /// the widget previously set on the [AnimatedChildSwitcher] as a child.
 ///
-/// More than one previous child can exist and be fading out while the newest
-/// one is fading in if they are swapped fast enough (i.e. before [duration]
-/// elapses).
+/// If they are swapped fast enough (i.e. before [duration] elapses), more than
+/// one previous child can exist and be transitioning out while the newest one
+/// is transitioning in.
+///
+/// If the "new" child is the same widget type as the "old" child, but with
+/// different parameters, then [AnimatedChildSwitcher] will *not* do a
+/// transition between them, since as far as the framework is concerned, they
+/// are the same widget, and the existing widget can be updated with the new
+/// parameters. If you wish to force the transition to occur, set a [Key]
+/// (typically a [UniqueKey] or [ValueKey], depending upon the desired behavior)
```

Comment:
unlikely that a UniqueKey is what one wants

Reply:
~~Oh? Why?  Isn't it going to generate a unique key each time, which is what the user might sometimes want here?  I can see cases where it might be onerous to collect all of the "values" and decide if the widget is different visually, and instead they just put a UniqueKey on it each time they swap it out.~~

Edit: I see from below why (fading when rotating, etc.).